agent/agent/src/main/java/com/microsoft/applicationinsights/agent/Main.java

```diff
@@ -0,0 +1,158 @@
+  public static void main(String[] args) throws Exception {
+    if (args.length == 1 && args[0].equals("unsign")) {
+      unsign();
+    } else {
+      System.err.println("Incorrect usage");
+    }
+  }
+
+  private static void unsign() throws Exception {
+    Path agentJar = getAgentJar();
+
+    Path tempJar = Files.createTempFile("applicationinsights-", ".jar");
+
+    byte[] buffer = new byte[1024];
+
+    try (JarInputStream in = new JarInputStream(Files.newInputStream(agentJar))) {
+
+      Manifest inManifest = in.getManifest();
+
+      if (!isSigned(inManifest)) {
+        System.err.println();
+        System.err.println("The agent jar is already unsigned");
+        System.exit(0);
+      }
+
+      Manifest outManifest = new Manifest();
+      outManifest.getMainAttributes().putAll(inManifest.getMainAttributes());
+
+      JarOutputStream out = new JarOutputStream(Files.newOutputStream(tempJar), outManifest);
+
+      JarEntry entry;
+      while ((entry = in.getNextJarEntry()) != null) {
+        if (SIGNATURE_FILE.matcher(entry.getName()).matches()) {
+          continue;
+        }
+        out.putNextEntry(entry);
+        int read;
+        while ((read = in.read(buffer)) != -1) {
+          out.write(buffer, 0, read);
+        }
+      }
+
+      out.close();
+    }
+
+    Path tempClassPath = Files.createTempDirectory("applicationinsights-");
+
+    String resourceName = Copy.class.getName().replace('.', '/') + ".class";
+    InputStream in = Main.class.getClassLoader().getResourceAsStream(resourceName);
+
+    if (in == null) {
+      System.err.println("Could not find resource: " + resourceName);
+      System.exit(1);
+    }
+
+    Path destinationPath = tempClassPath.resolve(resourceName);
+
+    Files.createDirectories(destinationPath.getParent());
+    Files.copy(in, destinationPath);
+
+    List<String> command =
+        asList(
+            getJavaExecutable(),
+            "-cp",
+            tempClassPath.toString(),
+            Copy.class.getName(),
+            tempJar.toString(),
+            agentJar.toString());
+
+    if (DEBUG) {
+      System.out.println(String.join(" ", command));
+    }
+
+    ProcessBuilder builder = new ProcessBuilder().command(command);
+
+    if (DEBUG) {
+      builder.redirectOutput(new File("unsign.out")).redirectError(new File("unsign.err"));
+    }
+
+    builder.start();
+  }
+
+  private static boolean isSigned(Manifest manifest) {
+    for (Map.Entry<String, Attributes> entry : manifest.getEntries().entrySet()) {
+      for (Object key : entry.getValue().keySet()) {
+        if (key instanceof Attributes.Name && key.toString().endsWith("-Digest")) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private static Path getAgentJar() throws URISyntaxException {
+    CodeSource codeSource = Main.class.getProtectionDomain().getCodeSource();
+
+    if (codeSource == null) {
+      throw new IllegalStateException("could not get agent jar location");
```

Comment:
do we want to sys.err gracefully here or truly bomb out?

Reply:
I think bombing is ok, I'm not sure why code source would be null, so not sure how to even give a more constructive error message (other than asking the user to open a github issue?)